03.bowling/bowling.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+def main
+  scoreboard = ARGV[0]
+  frames = split_by_frames(scoreboard)
+  final_score = calculate_score(frames)
+  output_result(final_score)
+end
+
+def strike?(frame)
+  frame == [10, 0]
+end
+
+def spare?(frame)
+  !strike?(frame) && (frame[0] + frame[1] == 10)
+end
+
+def groupe_last_frame_shots(frames, last_frame, last_frame_added_shot)
+  # 10フレーム目が全てストライクの場合
+  if strike?(last_frame) && strike?(last_frame_added_shot)
+    last_frame_final_shot = frames[11]
+    grouped_3_shots = (last_frame + last_frame_added_shot + last_frame_final_shot).reject(&:zero?)
+    frames[0..-4].push(grouped_3_shots)
+  else
+    grouped_3_shots = (last_frame + last_frame_added_shot).reject(&:zero?)
+    frames[0..-3].push(grouped_3_shots)
+  end
+end
+
+def split_by_frames(scoreboard)
+  scores = scoreboard.split(',')
+
+  shots = scores.flat_map do |score|
+    score == 'X' ? [10, 0] : score.to_i
+  end
+
+  frames = shots.each_slice(2).to_a
+  last_frame = frames[9]
+  last_frame_added_shot = frames[10]
+
+  # 最後のフレームがスペアかストライクの場合は3投目が投げられる。この後のcalculate_scoreで扱いやすいように最後のフレームを3shotsにまとめる
+  frames = groupe_last_frame_shots(frames, last_frame, last_frame_added_shot) if frames.size != 10
+  frames
+end
+
+def calculate_score(frames)
+  final_score = 0
+
+  frames.each_with_index do |frame, i|
+    final_score += frame.sum
+    next_frame = frames[i + 1]
+    after_next_frame = frames[i + 2]
+
+    break if i == frames.size - 1
+
+    if spare?(frame)
+      final_score += next_frame[0]
+    elsif strike?(frame)
+      final_score += if !strike?(next_frame)
+                       next_frame[0] + next_frame.fetch(1, 0)
```

Comment:
ちなみにここってfetchを使わないと問題になるケースってあります？
ないならこんな風に書いた方がすっきりしそうです。

```suggestion
                       next_frame[0..1].sum
```

もしくはこうとか？

```suggestion
                       next_frame.sum
```

Reply:
ご指摘ありがとうございます！
デバッグツールでも確認しましたが、確かにfetchする必要はないですね…！
(commit:[ec0970e](https://github.com/Judeeeee/ruby-practices/pull/5/commits/ec0970ee6a24e1bf8de754acc5ca07120a7abffa))
(以前、ストライクを[10,0]ではなく[10]として実装を進めていたのでその名残だと思います😓)

「ストライクのフレームの得点は次の2投の点を加算する」なので、`next_frame[0..1].sum`を採用しました〜！
(`next_frame.sum`だと最後のフレームが3投の場合に3投の合計になってしまう)